File: app/src/main/java/com/donald/musictheoryapp/Music/MusicXML/Notations.java
package com.donald.musictheoryapp.Music.MusicXML;

import org.json.JSONException;
import org.json.JSONObject;

public class Notations
{
	public static class NoteArrow
	{
		public final String label;

		public NoteArrow(String label)
		{
			this.label = label;
		}

		public NoteArrow(NoteArrow that)
		{
			this.label = that.label;
		}

		public boolean equals(NoteArrow that)
		{
			return that != null &&
				this.label.equals(that.label);
		}

		public static NoteArrow fromJSON(JSONObject object) throws JSONException
		{
			return new NoteArrow(
				object.getString("label")
			);
		}
	}

	public final NoteArrow noteArrow;

	public Notations(NoteArrow noteArrow)
	{
		this.noteArrow = noteArrow;
	}

	public Notations(Notations that)
	{
		this.noteArrow = new NoteArrow(that.noteArrow);
	}

	public boolean equals(Notations that)
	{
		return that != null &&
			(this.noteArrow == null ? that.noteArrow == null : this.noteArrow.equals(that.noteArrow));
	}

	public static Notations fromJSON(JSONObject object) throws JSONException
	{
		return new Notations(
			NoteArrow.fromJSON(object)
		);
	}

	public Notations clone()
	{
		return new Notations(noteArrow);
	}
}

File: app/src/main/java/com/donald/musictheoryapp/music/musicxml/Notations.kt
package com.donald.musictheoryapp.music.musicxml

import kotlin.Throws
import org.json.JSONException
import org.json.JSONObject
import org.dom4j.Element

data class Notations(
    val noteArrow: NoteArrow?
) {

    data class NoteArrow(
        val label: String
    ) {
        constructor(that: NoteArrow) : this(that.label)

        fun addToXml(notations: Element) {
            notations.addElement("other-notation").apply {
                addAttribute("notation-name", "note-arrow")
                addText(label)
            }
        }

        fun equals(that: NoteArrow?): Boolean {
            return that != null && label == that.label
        }

        @Throws(JSONException::class)
        fun toJson(): JSONObject {
            val `object` = JSONObject()
            `object`.put("label", label)
            return `object`
        }

        companion object {
            @Throws(JSONException::class)
            fun fromJson(`object`: JSONObject): NoteArrow {
                return NoteArrow(
                    `object`.getString("label")
                )
            }
        }

    }

    constructor(that: Notations) : this(that.noteArrow!!)

    fun equals(that: Notations?): Boolean {
        return that != null &&
            if (noteArrow == null) that.noteArrow == null else noteArrow.equals(that.noteArrow)
    }

    @Throws(JSONException::class)
    fun toJson(): JSONObject {
        val `object` = JSONObject()
        if (noteArrow != null) `object`.put("note_arrow", noteArrow.toJson())
        return `object`
    }

    fun addToXml(note: Element) {
        note.addElement("notations").apply {
            noteArrow?.addToXml(this)
        }
    }

    fun clone(): Notations {
        return Notations(noteArrow)
    }

    companion object {
        @JvmStatic
		@Throws(JSONException::class)
        fun fromJson(`object`: JSONObject): Notations {
            return Notations(
                NoteArrow.fromJson(`object`)
            )
        }
    }

}